Improve debuggability of activation checkpoint

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103859
* #103734

This PR makes some improvements for debuggability of checkpointing:
- improved error messages that are more understandable
- errors are now `CheckpointError` which subclasses `RuntimeError` (only `CheckpointError` triggers debug message, see below)
- stricter error checking by default:
   - shapes, dtypes, and device are compared 
   - we also now error when more tensors are being saved for backward during recompute
   - NOTE: checks are relaxed if it is detected that you are doing backward within forward
 - shapes, dtype, and device checking can be disabled by passing `determinism_check="none"`
 - new debug flag: more helpful error message when `debug=True`

Note:
- cpp stack trace is only included for x86 linux machines
- the error message if cpp stack trace is included can be quite long. For a function checkpointed with 8 operators, the log was around 1300 lines! (should this be hidden behind a flag?)

[Error message when debug='True' (python stack trace only)](https://gist.github.com/soulitzer/3d5e19c7cceae8e22f9bdd625ec39dd4)

[Error message when debug='True' (with python and cpp stacktrace)](https://gist.github.com/soulitzer/ff8fd8c3ccbb2c90dfe3df6d7713b167)

soulitzer committed:
Improve debuggability of activation checkpoint

[ghstack-poisoned]

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -290,6 +293,8 @@ def checkpoint(
     use_reentrant: Optional[bool] = None,
     context_fn: Callable[[], Tuple[ContextManager, ContextManager]] = noop_context_fn,
+    determinism_check: str = "default",
+    debug: bool = False,
```
Why two different options? Why not put the determism check in the debug config?

soulitzer replied:
We want to separate them because determinism_check="default" is something the user should keep on by default.
but debug=True (running logging_mode all the time) can impact perf.

albanD replied:
But you could have `debug="default"` and "full" and "none" for the other options? Or do you think that there are cases where people will want debug without determinism checks?

soulitzer replied:
> Or do you think that there are cases where people will want debug without determinism checks?

Ahh no, I'd say the only concern is if we wish to add more types of determinism_checks in the future, for example to check tensor statistics, it would be a little harder to do if the two options are coupled.

albanD replied:
We could also have a `debug="full_values"` ? The same way you can add things to determinism_check?

soulitzer replied:
Right, I think it makes sense for "full_values" to always have verbose logs. 

But a trickier example is if we want to provide the option for the user to specify their own determinism_check. Someone may have a check that may or may not be intended to be on all the time.

A possible UX today is to allow the user to pass in a function to `determinism_check`, but that may be more difficult if we only had the `debug` option.

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -405,16 +422,19 @@ def checkpoint(
 
     if use_reentrant:
-        if context_fn is not noop_context_fn:
+        if context_fn is not noop_context_fn or debug is not False:
```
Should you check determinism mode here?

soulitzer replied:
we should just update the docs to say "default" means nothing is checked for use_reentrant=True
I don't know if I want to force users to have to explicitly specify determinism_check="none" when using use_reentrant=True

albanD replied:
Ho that's because that's the default value. I would agree that we don't want to break all the current users yes :D 
I would say error on non none or default then yes.

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -735,6 +755,217 @@ def __init__(self, recompute_fn, early_stop):
+ Note that the logs can be quite long. Here's how they are structured:
+ (Tip: you can Ctrl-f for these headers)
+
+ 1. Stack traces of the operators that ran in the original forward
```
Follow up proposal: run the diff here and only print the stack traces for the diff.

albanD reviewed (commented):
Sounds good!
Only question about the flags and UX.

soulitzer committed:
Update on "Improve debuggability of activation checkpoint"


This PR makes some improvements for debuggability of checkpointing:
- improved error messages that are more understandable
- errors are now `CheckpointError` which subclasses `RuntimeError` (only `CheckpointError` triggers debug message, see below)
- stricter error checking by default:
   - shapes, dtypes, and device are compared 
   - we also now error when more tensors are being saved for backward during recompute
   - NOTE: checks are relaxed if it is detected that you are doing backward within forward
 - shapes, dtype, and device checking can be disabled by passing `determinism_check="none"`
 - new debug flag: more helpful error message when `debug=True`

Note:
- cpp stack trace is only included for x86 linux machines
- the error message if cpp stack trace is included can be quite long. For a function checkpointed with 8 operators, the log was around 1300 lines! (should this be hidden behind a flag?)

[Error message when debug='True' (python stack trace only)](https://gist.github.com/soulitzer/3d5e19c7cceae8e22f9bdd625ec39dd4)

[Error message when debug='True' (with python and cpp stacktrace)](https://gist.github.com/soulitzer/ff8fd8c3ccbb2c90dfe3df6d7713b167)

[ghstack-poisoned]

albanD reviewed (approved):
Sounds good to future proof this!

soulitzer commented:
@pytorchbot merge -f "Unrelated failures"
